rllib/evaluation/trajectory.py

```diff
@@ -0,0 +1,274 @@
+        # gets inserted.
+        self.cursor: int = 0
+        # The offset inside our buffer where the current trajectory starts.
+        self.trajectory_offset: int = 0
+        # The offset inside our buffer, from where to build the next
+        # SampleBatch.
+        self.sample_batch_offset: int = 0
+
+    @property
+    def timestep(self) -> int:
+        """The timestep in the (currently ongoing) trajectory."""
+        return self.cursor - self.trajectory_offset
+
+    def add_init_obs(self,
+                     env_id: EnvID,
+                     agent_id: AgentID,
+                     policy_id: PolicyID,
+                     init_obs: TensorType) -> None:
+        """Adds a single initial observation (after env.reset()) to the buffer.
+
+        Args:
+            env_id (EnvID): Unique id for the episode we are adding the initial
+                observation for.
+            agent_id (AgentID): Unique id for the agent we are adding the
+                initial observation for.
+            policy_id (PolicyID): Unique id for policy controlling the agent.
+            init_obs (TensorType): Initial observation (after env.reset()).
+        """
+        # Our buffer should be empty when we add the first observation.
+        if SampleBatch.CUR_OBS not in self.buffers:
+            assert self.has_initial_obs is False
+            assert self.cursor == self.sample_batch_offset == \
+                self.trajectory_offset == 0
+            self.has_initial_obs = True
+            # Build the buffer only for "obs" (needs +1 time step slot for the
+            # last observation). Only increase `self.timestep` once we get the
+            # other-than-obs data (which will include the next obs).
+            obs_buffer = np.zeros(
+                shape=(self.buffer_size + 1, ) + init_obs.shape,
+                dtype=init_obs.dtype)
+            obs_buffer[0] = init_obs
+            self.buffers[SampleBatch.CUR_OBS] = obs_buffer
+        else:
+            assert self.has_initial_obs
+            self.buffers[SampleBatch.CUR_OBS][self.cursor] = init_obs
+
+        self.env_id = env_id
+        self.agent_id = agent_id
+        self.policy_id = policy_id
+
+    def add_action_reward_next_obs(self,
+                                   env_id: EnvID,
+                                   agent_id: AgentID,
+                                   policy_id: PolicyID,
+                                   values: Dict[str, TensorType]) -> None:
+        """Add the given dictionary (row) of values to this batch.
+
+        Args:
+            env_id (EnvID): Unique id for the episode we are adding the initial
+                observation for.
+            agent_id (AgentID): Unique id for the agent we are adding the
+                initial observation for.
+            policy_id (PolicyID): Unique id for policy controlling the agent.
+            values (Dict[str, TensorType]): Data dict (interpreted as a single
```

Comment:
Why not pass these separately?

Reply:
As **kwargs? It's so much more slower, if I remember correctly:
The call to `add_values/add_action_reward_next_obs` is the most expensive one in the whole sampler loop, by far.

```
import time

num = 1000000

def f(d):
    return

def f_kwargs(**d):
    return

d_in = {str(i): 0 for i in range(100)}
start = time.time_ns()
for _ in range(num):
    f(d_in)
delta = time.time_ns() - start
print("delta / num={}".format(delta / num))

start = time.time_ns()
for _ in range(num):
    f_kwargs(**d_in)
delta = time.time_ns() - start
print("delta / num={}".format(delta / num))
```

Reply:
Yup, output:
```
delta / num=124.0005
delta / num=3793.9991
```